### ISSUE TITLE
Does not play nicely with Enupal Stripe Payments

### ISSUE THREAD

**Issue by MattWilcox** [putyourlightson/craft-snaptcha#23]
Hi Ben,

looks like there's some issue with Snaptcha where it is stopping Enupal Stripe Payments from having a form that actually works. It's been an issue we've had for a while, but Andre managed to track it down to this rather than an issue with the Enupal plugin itself:

```
if you disable the Snaptcha plugin you may notice that the payment form starts working again. 
For some reason the plugin is blocking our ajax call to create the Stripe Session
and returning this error:

JavaScript must be enabled for your submission to be verified. Please enable it and refresh the page.
```

Any help greatly appreciated as we've had to disable SCA payments on a live site for some time while this has been happening. Is this a configuration issue on our end, or a genuine bug?

We can supply a test site and link with the issue in-play if that's any help. 

https://github.com/enupal/stripe/issues/217 - Issue on their plugin, though it's missing some direct-email back and forth.

**Comment by bencroker** [comment]
There is a config setting for adding controller actions to exclude from validation. Can you try excluding the controller action that is being blocked?

```php
'excludeControllerActions' => ['enupal/stripe/create-checkout-session'],
```

https://github.com/putyourlightson/craft-snaptcha/blob/62bceba534f38060a0d2ba847cebc826247eca06/src/config.php#L55-L56

**Comment by MattWilcox** [comment]
That file was missing, that's our bad.
With that said, even with that file now existing and that line in there, this still isn't working properly and is returning:

```
(()=>{var e,n={498:()=>{var e={};!function(n){"use strict";e={paymentFormsList:{},finalData:{},zeroDecimals:{},init:function(){this.paymentFormsList=n(".enupal-stripe-form"),this.zeroDecimals=["MGA","BIF","CLP","PYG","DJF","RWF","GNF","UGX","JPY","VND","VUV","XAF","KMF","KRW","XOF","XPF"],this.paymentFormsList.each((function(){var t=n(this);e.initializeForm(t),n(this).keydown((function(e){if(13===e.keyCode)return e.preventDefault(),!1}))}))},initializeForm:function(t){if(void 0===n(t).find('[name="enupalStripe[stripeData]"]').val())return!1;var a=n.parseJSON(n(t).find('[name="enupalStripe[stripeData]"]').val());a.lineItems=t.find('[name="enupalStripe[enupalLineItems]"]').val(),a.allowPromotionCodes=t.find('[name="enupalStripe[enupalAllowPromotionCodes]"]').val(),a.removeDefaultItem=t.find('[name="enupalStripe[enupalRemoveDefaultItem]"]').val(),navigator.userAgent.indexOf("Firefox")<0&&(n(t).find('[name="enupalStripe[stripeData]"]').val(""),t.find('[name="enupalStripe[enupalLineItems]"]').val(""),t.find('[name="enupalStripe[enupalAllowPromotionCodes]"]').val(""),t.find('[name="enupalStripe[enupalRemoveDefaultItem]"]').val("")),this.finalData.finalAmount=a.stripe.amount;var i=null,o=this,s="stripe-payments-submit-button-"+a.paymentFormId;if(!a.useSca){i=StripeCheckout.configure({key:a.pbk,token:function(n,i){t.find('[name="enupalStripe[token]"]').val(n.id),t.find('[name="enupalStripe[email]"]').val(n.email),e.addValuesToForm(t,i,a),o.showProcessingText(t,a),t.unbind("submit",[t,a]),t.submit()},opened:function(){},closed:function(){}})}if(a.coupon.enabled){var r=n("#check-coupon-button-"+a.paymentFormId);this.updateTotalAmoutLabel(t,a);var u=t.find("#remove-coupon-"+a.paymentFormId);this.updatesTotalLabelOnChoosesPlan(t,a),u.click((function(e){e.preventDefault(),o.handleRemoveCoupon(t,a,o)})),r.click((function(e){e.preventDefault(),o.handleCouponValidation(t,a,o)}))}t.find("#"+s).on("click",(function(n){var o=t[0];o.checkValidity()?(n.preventDefault(),e.submitPayment(t,a,i)):o.reportValidity&&o.reportValidity()}))},showProcessingText:function(e,n){e.find("#stripe-payments-submit-button-"+n.paymentFormId).prop("disabled",!0),e.find("#stripe-payments-submit-button-"+n.paymentFormId).find("span").length?e.find("#stripe-payments-submit-button-"+n.paymentFormId).find("span").text(n.paymentButtonProcessingText):e.find("#stripe-payments-submit-button-"+n.paymentFormId).text(n.paymentButtonProcessingText)},updateTotalAmoutLabel:function(e,n){var t=e.find("#total-amount-value-"+n.paymentFormId),a=this.getFinalAmount(e,n);t&&t.text(a)},updatesTotalLabelOnChoosesPlan:function(e,t){var a=this;if(0!==t.subscriptionType)if("radio"==t.subscriptionStyle){var i=e.find('input[name="enupalStripe[enupalMultiPlan]"]');n(i).change((function(){n(this).is(":checked")&&a.handleRemoveCoupon(e,t,a)}))}else{var o=e.find('[name="enupalStripe[enupalMultiPlan]"]');n(o).on("change",(function(){a.handleRemoveCoupon(e,t,a)}))}},handleRemoveCoupon:function(e,n,t){var a=e.find("#coupon-message-"+n.paymentFormId),i=e.find("#remove-coupon-"+n.paymentFormId),o=e.find("#couponCode-"+n.paymentFormId);o&&o.val(""),a&&a.text(""),i.addClass("enupal-hidden"),e.find('[name="enupalCouponCode"]').val(""),t.updateTotalAmoutLabel(e,n)},handleCouponValidation:function(e,t,a){var i=e.find("#check-coupon-button-"+t.paymentFormId),o=e.find("#coupon-message-"+t.paymentFormId),s=e.find("#remove-coupon-"+t.paymentFormId),r=e.find("#total-amount-value-"+t.paymentFormId),u=e.find("#couponCode-"+t.paymentFormId),l=u.val(),p=t.stripe;u.val("");var d=this.convertToCents(a.getFinalAmount(e,t),p.currency);i.prop("disabled",!0);var c={action:"/enupal-stripe/coupons/validate",amount:d,couponCode:l,isRecurring:this.getIsRecurring(e,t),currency:p.currency,successMessage:t.coupon.successMessage};n.ajax({type:"POST",url:"enupal/validate-coupon",data:c,dataType:"json",success:function(n){!0===n.success?(o.removeClass("coupon-error"),o.text(n.successMessage),s.removeClass("enupal-hidden"),r&&r.text(n.finalAmount),e.find('[name="enupalCouponCode"]').val(n.coupon.id)):(o.addClass("coupon-error"),s.addClass("enupal-hidden"),o.text(t.coupon.errorMessage)),i.prop("disabled",!1)}.bind(this),error:function(e,n,t){i.prop("disabled",!1),console.error(e,n,t.toString())}.bind(this)})},getIsRecurring:function(e,n){var t=!1;return 1==n.amountType&&(t=e.find('[name="enupalStripe[recurringToggle]"]').is(":checked")),n.enableSubscriptions&&(t=!0),t},addValuesToForm:function(e,n,t){if(t.stripe.shippingAddress){var a="shipping",i="address";this.setAddressToHiddenValues(a,i,n,e)}if(t.stripe.billingAddress){a="billing",i="billingAddress";this.setAddressToHiddenValues(a,i,n,e)}},setAddressToHiddenValues:function(e,n,t,a){t[e+"_name"]&&a.find('[name="enupalStripe['+n+'][name]"]').val(t[e+"_name"]),t[e+"_address_country"]&&a.find('[name="enupalStripe['+n+'][country]"]').val(t[e+"_address_country_code"]),t[e+"_address_zip"]&&a.find('[name="enupalStripe['+n+'][zip]"]').val(t[e+"_address_zip"]),t[e+"_address_state"]&&a.find('[name="enupalStripe['+n+'][state]"]').val(t[e+"_address_state"]),t[e+"_address_line1"]&&a.find('[name="enupalStripe['+n+'][line1]"]').val(t[e+"_address_line1"]),t[e+"_address_city"]&&a.find('[name="enupalStripe['+n+'][city]"]').val(t[e+"_address_city"])},redirectToCheckoutSession:function(e,t){this.showProcessingText(e,t);var a=e.serializeArray();a.push({name:"enupalStripe[enupalLineItems]",value:t.lineItems}),a.push({name:"enupalStripe[enupalAllowPromotionCodes]",value:t.allowPromotionCodes}),a.push({name:"enupalStripe[enupalRemoveDefaultItem]",value:t.removeDefaultItem}),a.push({name:"action",value:"/enupal-stripe/checkout/create-session"}),a.push({name:"enupalStripeData",value:JSON.stringify(t)}),n.ajax({type:"POST",url:"enupal/stripe/create-checkout-session",data:a,dataType:"json",success:function(e){!0===e.success?Stripe(t.pbk).redirectToCheckout({sessionId:e.sessionId}):console.log("Something went wrong, please contact the admin")}.bind(this),error:function(e,n,t){console.error(e,n,t.toString())}.bind(this)})},submitPayment:function(e,t,a){var i=n.extend(!0,{},t),o=this,s=i.stripe;if(s.amount=this.convertToCents(this.getFinalAmount(e,i),s.currency),e.find('[name="enupalStripe[amount]"]').val(s.amount),e.find('[name="enupalStripe[testMode]"]').val(i.testMode),t.coupon.enabled){var r=e.find('[name="enupalCouponCode"]').val(),u=this.getIsRecurring(e,t);if(r){var l={action:"/enupal-stripe/coupons/validate",amount:s.amount,couponCode:r,isRecurring:u,currency:s.currency,successMessage:t.coupon.successMessage};n.ajax({type:"POST",url:"enupal/validate-coupon",data:l,dataType:"json",success:function(n){!0===n.success&&(s.amount=n.finalAmountInCents),t.useSca?o.redirectToCheckoutSession(e,i):a.open(s)}.bind(this),error:function(e,n,t){console.error(e,n,t.toString())}.bind(this)})}else t.useSca?this.redirectToCheckoutSession(e,i):a.open(s)}else t.useSca?this.redirectToCheckoutSession(e,i):a.open(s)},getFinalAmount:function(e,t){var a=t.stripe.amount,i=0,o=!1;if(t.enableSubscriptions)if(0==t.subscriptionType){if(t.singleSetupFee>0&&(i=t.singleSetupFee),t.enableCustomPlanAmount)"undefined"!==(r=e.find('[name="enupalStripe[customPlanAmount]"]').val())&&r>0&&(a=r)}else{var s=null,r=null;if((s="radio"==t.subscriptionStyle?n('input[name="enupalStripe[enupalMultiPlan]"]:checked').val():e.find('[name="enupalStripe[enupalMultiPlan]"]').val())in t.multiplePlansAmounts&&(r=t.multiplePlansAmounts[s].amount,t.stripe.currency=t.multiplePlansAmounts[s].currency),s in t.setupFees){var u=t.setupFees[s];u>0&&(i=u)}"undefined"!==r&&r>0&&(a=r)}else if(1==t.amountType){var l=e.find('[name="enupalStripe[customAmount]"]').val();o=e.find('[name="enupalStripe[recurringToggle]"]').is(":checked"),"undefined"!==l&&l>0&&(a=l)}if((t.applyTax||o)&&t.enableTaxes){var p=parseFloat(t.tax/100*(parseFloat(a)+parseFloat(i))).toFixed(2);a=parseFloat(a)+parseFloat(p);var d=t.taxLabel+": "+t.currencySymbol+p;e.find('[name="enupalStripe[taxAmount]"]').val(p),e.find('[name="tax-amount-label"]').empty().append(d)}return parseFloat(a)+parseFloat(i)},convertToCents:function(e,n){return this.hasZeroDecimals(n)?e:(100*e).toFixed(0)},convertFromCents:function(e,n){return this.hasZeroDecimals(n)?e:e/100},hasZeroDecimals:function(e){for(var n=0;n<this.zeroDecimals.length;n++)if(this.zeroDecimals[n]===e.toUpperCase())return!0;return!1}},n(document).ready((function(n){e.init()}))}(jQuery)},615:()=>{},678:()=>{}},t={};function a(e){var i=t[e];if(void 0!==i)return i.exports;var o=t[e]={exports:{}};return n[e](o,o.exports,a),o.exports}a.m=n,e=[],a.O=(n,t,i,o)=>{if(!t){var s=1/0;for(p=0;p<e.length;p++){for(var[t,i,o]=e[p],r=!0,u=0;u<t.length;u++)(!1&o||s>=o)&&Object.keys(a.O).every((e=>a.O[e](t[u])))?t.splice(u--,1):(r=!1,o<s&&(s=o));if(r){e.splice(p--,1);var l=i();void 0!==l&&(n=l)}}return n}o=o||0;for(var p=e.length;p>0&&e[p-1][2]>o;p--)e[p]=e[p-1];e[p]=[t,i,o]},a.o=(e,n)=>Object.prototype.hasOwnProperty.call(e,n),(()=>{var e={815:0,984:0,879:0};a.O.j=n=>0===e[n];var n=(n,t)=>{var i,o,[s,r,u]=t,l=0;if(s.some((n=>0!==e[n]))){for(i in r)a.o(r,i)&&(a.m[i]=r[i]);if(u)var p=u(a)}for(n&&n(t);l<s.length;l++)o=s[l],a.o(e,o)&&e[o]&&e[o][0](),e[s[l]]=0;return a.O(p)},t=self.webpackChunkstripe=self.webpackChunkstripe||[];t.forEach(n.bind(null,0)),t.push=n.bind(null,t.push.bind(t))})(),a.O(void 0,[984,879],(()=>a(498))),a.O(void 0,[984,879],(()=>a(615)));var i=a.O(void 0,[984,879],(()=>a(678)));i=a.O(i)})();
```

It's late here so I shant be looking at this again until tomorrow - but cheers!

**Comment by bencroker** [comment]
Be sure to check the response body, if Snaptcha is blocking the request then that should be reflected in an identifiable error message.

**Comment by MattWilcox** [comment]
Boom, now working though the setting is `'excludeControllerActions' => ['enupal-stripe/checkout/create-session']` which tripped me up.

Thanks Ben, that's been A Journey to work out what's been wrong.

**Comment by bencroker** [comment]
Great, added in https://github.com/putyourlightson/craft-snaptcha/commit/5240f9899acb9414c62ffd4bea6e6f53f33350b7 and released in version 3.0.10 to ensure this doesn't affect other sites. 
Thanks for reporting this Matt!
